rp2/boards: Add WEACTSTUDIO boards with BOARD_VARIANT to select the flash size

mattytrentini opened the pull request:
This is adding a board definition for the [Wavgat Pico Board RP2040](https://www.aliexpress.com/item/1005003843712530.html). This is for the 16MB version; once this PR is in good shape I'll create the 2/4/8 MB versions too.

Note that `lib/mbedtls` has been added as a submodule; this has _nothing to do with this board_ - it appears to be necessary for the rp2 port?

Also of note: I'm not aware of any VID/PID for the Wavgat boards so I left these defines out. It seems to work but I'm not sure that's the right thing to do...?

Finally, it was somewhat unexpected that pico-sdk typically has [definitions for each supported board](https://github.com/raspberrypi/pico-sdk/tree/master/src/boards/include/boards). This is the first MicroPython example of a board that is defined _outside_ of the pico-sdk repository (by specifying `PICO_BOARD_HEADER_DIRS` and creating `wavgat_16mb.h` locally).

mattytrentini committed:
rp2/boards/WEACTSTUDIO: Add WEACTSTUDIO with multiple variants.

This supports 2, 4, 8 and 16MB flash variants.

mcauser commented:
I thought Wavgat was more of a reseller than a board fab.
Here's the same board from RQG: https://www.aliexpress.com/item/1005003944981647.html

The form factor reminds me of a WeAct back-pill board, eg: https://www.aliexpress.com/item/1005003658370449.html

Found it on TaoBao: https://item.taobao.com/item.htm?id=671510282412
Which links to: https://gitee.com/WeAct-TC/rp2040
Which labels the board: **WeAct-RP2040-V1.0**
https://github.com/WeActTC/WeActStudio.RP2040CoreBoard

It's a WeAct board.

mcauser commented:
_... they do_

![O1CN01g9fYqx1R49RBIrN9e_!!781802057](https://user-images.githubusercontent.com/1038959/162385169-058fbaa9-4568-4d6d-b565-743a3bd84b4f.png)

Looks like this Wavgat board is a clone. In their back side of the board image you can see where they blacked out the WeAct logo:

<img width="364" alt="Screen Shot 2022-04-08 at 5 24 20 pm" src="https://user-images.githubusercontent.com/1038959/162385350-1fc1a9f6-379d-471d-9ac8-e8ee5b2f6a06.png">

mattytrentini commented:
> It's a WeAct board. 

So should I rename the board to WeAct? WeActStudio? Any suggestions?
I think the definition would remain intact (except the links to Wavgat should be removed).

mattytrentini commented:
OK, renamed to WeAct Studio RP2040. Thanks @mcauser!

andypiper commented:
FWIW, I'm currently seeing a build error with this PR:

```text
[  9%] Generating frozen_content.c
MPY uasyncio/core.py
MPY uasyncio/event.py
MPY uasyncio/funcs.py
MPY uasyncio/lock.py
MPY board.py
error freezing mpy ['/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/_boot_fat.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/rp2.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/_boot.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/onewire.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/ds18x20.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/dht.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/uasyncio/__init__.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/uasyncio/core.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/uasyncio/event.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/uasyncio/funcs.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/uasyncio/lock.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/uasyncio/stream.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/neopixel.mpy', '/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/board.mpy']:
/Users/andypiper/Development/third-party/mpy-build/micropython/ports/rp2/boards/build_WEACTSTUDIO_16MB/frozen_mpy/rp2.mpy: corrupt .mpy file

make[2]: *** [frozen_content.c] Error 1
make[1]: *** [CMakeFiles/firmware.dir/all] Error 2
make: *** [all] Error 2
```

Note that I'm a n00b at building my own firmwares so this could be user error, but I've successfully built a base rp2040, and also a `PIMORONI_TINY2040` image.

mattytrentini commented:
Rebased and tidied up. 

It would be good to support the different storage options; should this be done with an optional build parameter? Is there any precedent for this? It doesn't seem to make sense to have completely separate board definitions when the only thing that changes is the flash size...

mattytrentini committed:
rp2/Makefile: Add support for BOARD_VARIANTS.

Following stm32.  This allows a single board definition to define variants
of its configuration.

jimmo commented on ports/rp2/Makefile:
```diff
@@ -36,3 +40,8 @@ submodules:
 	$(MAKE) -f ../../py/mkrules.mk GIT_SUBMODULES="$${GIT_SUBMODULES}" submodules
+
+query-variants:
+	@BOARD_VARIANTS=$$(cmake -B $(BUILD) -DECHO_BOARD_VARIANTS=1 ${CMAKE_ARGS} -S . 2>&1 | \
```
It needs to use `-B $(BUILD)/variants` or something to force a different directory to the regular build (see how the `submodules` target uses `-B $(BUILD)/submodules`).

mattytrentini replied:
Done.

jimmo commented on ports/rp2/boards/WEACTSTUDIO/mpconfigboard.cmake:
```diff
@@ -0,0 +1,43 @@
+    math(EXPR PICO_FLASH_SIZE_BYTES "${FLASH_SIZE_MB} * 1024 * 1024")
+    add_compile_definitions(
+        PICO_FLASH_SIZE_BYTES=${PICO_FLASH_SIZE_BYTES}
+        FLASH_SIZE_MB=${FLASH_SIZE_MB}
```
Remove setting definition for `FLASH_SIZE_MB` here.

jimmo commented on ports/rp2/boards/WEACTSTUDIO/mpconfigboard.h:
```diff
@@ -0,0 +1,4 @@
+#define MICROPY_HW_BOARD_NAME "WeAct Studio RP2040"
+
+// Allow 1MB for the firmware image itself, allocate the remainder to the filesystem
+#define MICROPY_HW_FLASH_STORAGE_BYTES ((FLASH_SIZE_MB - 1) * 1024 * 1024)
```
Use PICO_FLASH_SIZE_BYTES instead

```
#define MICROPY_HW_FLASH_STORAGE_BYTES (PICO_FLASH_SIZE_BYTES - 1024 * 1024)
```

jimmo commented on ports/rp2/boards/WEACTSTUDIO/mpconfigboard.cmake:
```diff
@@ -0,0 +1,43 @@
+# Provide different variants for the downloads page.
+set(BOARD_VARIANTS "flash_2mb flash_4mb flash_8mb flash_16mb")
+
+if("${BOARD_VARIANT}" STREQUAL "flash_2mb")
```
Which one is the default variant? I think you'll end up building PICO?

mattytrentini replied:
I chose `flash_16mb` as the default; it seems like it'd be the most popular.

jimmo commented on ports/rp2/boards/WEACTSTUDIO/board.json:
```diff
@@ -0,0 +1,18 @@
+    "product": "WeAct Studio RP2040",
+    "url": "https://github.com/WeActTC/WeActStudio.RP2040CoreBoard/tree/master",
+    "vendor": "WeAct"
+}
```
Need to add variants here (but not the default variant).

jimmo commented on ports/rp2/boards/WEACTSTUDIO/weactstudio_16mb.h:
```diff
@@ -0,0 +1,73 @@
+// A pico-sdk board definition is required since the Wavgat boards are not officially
```
Can we make `weactstudion_common.h` and have them all share that except for setting `PICO_FLASH_SIZE_BYTES`.

mattytrentini replied:
Yes, though I just called the common file `weactstudio.h`.

dpgeorge commented:
I reorganised the commits into two separate ones and force pushed to this PR's branch.  The code is unchanged.
